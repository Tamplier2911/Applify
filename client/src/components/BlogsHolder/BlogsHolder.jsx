// import './BlogsHolder.scss';
import React, { useEffect, useState } from "react";
import { withRouter } from "react-router-dom";

// redux
import { connect } from "react-redux";
import { createStructuredSelector } from "reselect";
import { selectCurrentLanguage } from "../../redux/lang/lang.selectors";
import { loadAllBlogpostsStart } from "../../redux/blogs/blogs.actions";
import {
  selectAllBlogposts,
  selectIsLoading
} from "../../redux/blogs/blogs.selectors";

// components
import WithSpinnerHOC from "../WithSpinnerHOC/WithSpinnerHOC";
import CollectionHolder from "../CollectionHolder/CollectionHolder";
import CollectionForHolder from "../CollectionForHolder/CollectionForHolder";
import BlogCard from "../BlogCard/BlogCard";
import Button from "../Button/Button";
import FormInput from "../FormInput/FormInput";

// data formaters
import { simpleBlogsSearch } from "../../utils/DataTransformations/simpleSearches";

// JS Rendering CSS
import {
  BlogsHolderContainer,
  BlogsHolderInputBarContainer
} from "./BlogsHolderStyles";

// component constants
import blogsHolderData from "../../utils/ComponentBlogsHolderConstants/componentBlogsHolderConstants";

// buff collection for holder with spinner
const CollectionForHolderWithSpinner = WithSpinnerHOC(CollectionForHolder);

const BlogsHolder = ({
  lang,
  history,
  loadAllBlogpostsStart,
  blogposts,
  isLoading
}) => {
  useEffect(() => {
    if (!blogposts.length) {
      loadAllBlogpostsStart();
    }
  }, [loadAllBlogpostsStart, blogposts.length]);

  const [searchInput, setSearchInput] = useState({ search: "" });
  const { search } = searchInput;
  const onInputChange = e => {
    const { name, value } = e.target;
    setSearchInput({ ...searchInput, [name]: value });
  };

  const redirectToCreateBlogpost = () => {
    history.push("/profile/blogs/create");
  };

  const blogsArray = simpleBlogsSearch(blogposts, search.toLowerCase());

  const {
    blogsHolderTitle,
    blogsHolderButton,
    blogsHolderSearch
  } = blogsHolderData[lang];

  return (
    <BlogsHolderContainer>
      <Button
        type="button"
        value={blogsHolderButton}
        click={redirectToCreateBlogpost}
      />
      <BlogsHolderInputBarContainer>
        <FormInput
          onInputChange={e => onInputChange(e)}
          value={search}
          name="search"
          label={blogsHolderSearch}
          type="text"
        />
      </BlogsHolderInputBarContainer>
      <CollectionHolder
        lang={lang}
        title={blogsHolderTitle}
        refresher={1}
        cb={() => loadAllBlogpostsStart()}
      >
        <CollectionForHolderWithSpinner
          isLoading={isLoading}
          dataCollection={blogsArray}
        >
          <BlogCard />
        </CollectionForHolderWithSpinner>
      </CollectionHolder>
    </BlogsHolderContainer>
  );
};

const mapStateToProps = createStructuredSelector({
  lang: selectCurrentLanguage,
  blogposts: selectAllBlogposts,
  isLoading: selectIsLoading
});

export default withRouter(
  connect(mapStateToProps, { loadAllBlogpostsStart })(BlogsHolder)
);
