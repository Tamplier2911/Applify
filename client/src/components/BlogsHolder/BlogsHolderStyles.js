import styled from "styled-components";

export const BlogsHolderContainer = styled.div``;
