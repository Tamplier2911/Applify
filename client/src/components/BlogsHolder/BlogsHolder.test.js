import React from "react";

// component
import BlogsHolder from "./BlogsHolder";

// utils
import {
  setShallowWRC,
  findByTestAttr,
  storeFactory,
} from "../../tests/testUtils";

describe("<BlogsHolder />", () => {
  const initialStore = {
    lang: {
      languagePanelHidden: true,
      currentLanguage: "eng",
    },
    blogs: {
      allBlogposts: [],
      allSets: {},
      currentSet: [],
      currentSlot: 0,
      allSlots: 0,
      isLoading: false,
      errorMessage: null,
    },
  };
  const store = storeFactory(initialStore);

  let wrapper;
  beforeEach(() => {
    wrapper = setShallowWRC(<BlogsHolder store={store} />);
  });

  it("renders without an error", () => {
    const blogsHolder = findByTestAttr(wrapper, "blogs-holder");
    expect(blogsHolder.length).toBe(1);
  });

  it("renders child - blogs holder search", () => {
    const blogsHolderSearch = findByTestAttr(wrapper, "blogs-holder-search");
    expect(blogsHolderSearch.length).toBe(1);
  });

  it("renders child - blogs holder collection", () => {
    const blogsHolderCollection = findByTestAttr(
      wrapper,
      "blogs-holder-collection"
    );
    expect(blogsHolderCollection.length).toBe(1);
  });

  it("matches snapshot ", () => {
    const blogsHolder = findByTestAttr(wrapper, "blogs-holder");
    expect(blogsHolder.length).toMatchSnapshot();
  });
});
