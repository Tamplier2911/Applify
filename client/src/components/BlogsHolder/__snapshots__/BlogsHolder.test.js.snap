// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BlogsHolder /> matches snapshot  1`] = `1`;
